target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_arcadyan_vgv952cjw33-e-ir.dtsi

```diff
@@ -0,0 +1,409 @@
+				reg = <0x4640000 0xc0000>;
+			};
+
+			partition@4700000 {
+				label = "lcdimage";
+				reg = <0x4700000 0x300000>;
+			};
+
+			partition@4A00000 {
+				label = "mfgconfig";
+				reg = <0x4a00000 0x100000>;
+			};
+
+			partition@4B00000 {
+				label = "sipdata";
+				reg = <0x4b00000 0x100000>;
+			};
+
+			partition@4C00000 {
+				label = "voice";
+				reg = <0x4c00000 0x4000000>;
+			};
+
+			partition@8C00000 {			/* Called "misc" in original u-boot env. */
+				label = "ubi";			/* Rename to "ubi" for auto ubi attachment and usage of rootfs, rootfs_data, kernel volumes */
+				reg = <0x8c00000 0x13200000>;	/* Rename to "firmware" for special squashfs-rootfs/jffs2-rootfs_data treatment */
+			};
+
+			partition@1BE00000 {
+				label = "rootfs2";
+				reg = <0x1be00000 0x3c00000>;
+			};
+
+			partition@1FA00000 {
+				label = "kernel2";
+				reg = <0x1fa00000 0x500000>;
+			};
+
+			partition@1FF00000 {
+				label = "mystery";		/* Missing in original u-boot environment, seems to be empty (erased) */
+				reg = <0x1ff00000 0x100000>;
+			};
+		};
+	};
+
+	display@2 {					/* @2 because using ebu BUSCON2/ADDSEL2. Name not 'easybox904-display', */
+		compatible = "ilitek,ili9341_eb904";	/* see 'Node Names' at https://elinux.org/Device_Tree_Usage */
+		reg = <2 0x0 0x4>;			/* Size 0x4 contains 2 byte data and 2 byte command registers */
+		status = "okay";
+		rotate = <270>;
+		fps = <30>;
+		bgr;
+		buswidth = <8>;
+		reset-gpios = <&gpio 6 GPIO_ACTIVE_HIGH>;
+		led-gpios = <&gpio 28 GPIO_ACTIVE_HIGH>;
+		debug = <1>;
+	};
+};
+
+/*
+*  Following is the very experimental part which is broken and requires much more elaboration
+*/
+
+/* From FRITZ3370.dts */ 
+&gpio {
+	pinctrl-names = "default";
+	pinctrl-0 = <&state_default>;
+
+	state_default: pinmux {
+		phy-rst { /* I have no idea wether this makes sense */
+			lantiq,pins = "io37", "io44";			/* FRITZ3370.dts */
+		     /*	lantiq,pins = "io42";				// TDW8970.dts */
+			lantiq,pull = <0>;
+			lantiq,open-drain = <0>;
+			lantiq,output = <1>;
+						};
+
+		pcie-rst {
+			lantiq,pins = "io38";	/* FRITZ3370.dts, TDW8970.dts, EASY80920.dtsi, P2812HNUFX.dtsi */
+		     /*	lantiq,pins = "io21";	// ARV7519RW22.dts */
+			lantiq,pull = <0>;
+			lantiq,output = <1>;
+		};
+
+		exin1 {
+			lantiq,groups = "exin1";
+			lantiq,function = "exin";
+		};
+
+		conf_tp {
+			lantiq,pins = "io1"; /* exin1 */
+			lantiq,open-drain;
+			lantiq,pull = <0>;
+		};
+
+		conf_spi {
+			lantiq,pins = "io29", "io30", "io39"; /* gpiois for spi */
+			lantiq,pull = <2>; /*Pull Up*/
+		};
+	};
+};
+
+&usb_phy0 {
+	status = "okay";
+};
+
+&usb_phy1 {
+	status = "okay";
+};
+
+&usb0 {
+	status = "okay";
+	vbus-supply = <&usb_vbus>;
+};
+
+&usb1 {
+	status = "okay";
+	vbus-supply = <&usb_vbus>;
+};
+
+&eth0 {
+	interface@0 {
+		compatible = "lantiq,xrx200-pdi";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0>;
+		mac-address = [ 00 11 22 33 44 55 ];
+		lantiq,switch;
+		
+		ethernet@0 { /*rtl8367*/
```

Comment:
I tried to convert all this to DSA. But i did not get a single NIC let alone Wifi interface.

How to best come up with a description from scratch with the information others already found?

Can i start with one phy and add the others later? Must be something with `swconfig`, maybe check link status, list ports etc.

Looking at https://github.com/openwrt/openwrt/pull/3085/commits/91cd89475c81f978fdc97fd8c55e3fa233c82577 i would ask @xdarklight for some advise.

i do have

```
+&gswip {
+       pinctrl-0 = <&mdio_pins>;
+       pinctrl-names = "default";
+};
+&gswip_mdio {
+        phy0: ethernet-phy@0 {
+                reg = <0x0>;
+        };
+        phy4: ethernet-phy@4 {
+                reg = <0x4>;
+        };
+        phy5: ethernet-phy@5 {
+                reg = <0x5>;
+        };
+        phy13: ethernet-phy@13 {
+                reg = <0x13>;
+        };
+};
+&gswip_ports {
+        port@0 {
+                reg = <0>;
+                label = "lan0";
+                phy-mode = "rgmii";
+                phy-handle = <&phy0>;
+        };
+        port@1 {
+                reg = <1>;
+                label = "lan1";
+                phy-mode = "rgmii";
+                phy-handle = <&phy4>;
+        };
+        port@2 {
+                reg = <2>;
+                label = "lan2";
+                phy-mode = "rgmii";
+                phy-handle = <&phy5>;
+        };
+        port@3 {
+                reg = <3>;
+                label = "lan3";
+                phy-mode = "rgmii";
+                phy-handle = <&phy13>;
+        };
```

Reply:
here you are trying to use DSA with the built-in switch of the Lantiq SoC. However, this device has an additional switch RealTek IC, and for that there isn't yet a working DSA driver.

Reply:
It seems that drivers/net/dsa/realtek/ does exist since 5.18. Not sure what that is but it might just be what was missing so far.

Reply:
@luizluca

Reply:
The `rtl8365mb.c` does support some similar RealTek chips:
 *  - RTL8367RB-VB
 *  - RTL8367SB
 *  - RTL8367S

But RTL8367**B** is not in that list, but probably not much is missing in order to support it. The driver by now supports both, RealTek SMI and MDIO, however, comparing the OpenWrt swconfig drivers for the two chips (`diff -u rtl8367b.c rtl8367.c`) reveals quite major differences.

Reply:
i will maybe try `.compatible = "realtek,rtl8365mb"` and see if it probes or what might be needed on top

Reply:
> @luizluca

I'm here.

We do have an upstream Realtek DSA driver for some variants in rtl8366rb.c and rtl8365mb.c. These reference just the name of the version that was first supported. You can identify the "family" of that switch if you have access to the GPL code for that device (it would be the directory name). The rtl8365mb.c, for example, is related to driver rtl8367c and it can easily support any of these versions:

RTL8363NB
RTL8364NB
RTL8365MB-VC
RTL8367RB-VB
RTL8367SB
RTL8367S
RTL8366SC
RTL8363SC
RTL8370MB
RTL8310SR
RTL8363NB-VB
RTL8364NB-VB
RTL8363SC-VB

According to your DTS, you are using driver "rtl8367b". It is very close to the rtl8367c "family" and just recently it got support for rtl8367c switches (4b81eda3c1a9b9ef32640df9402e3650fb48d98b). It is a good sign that rtl8365mb.c could work with your device.

So, the first step is to really identify your chip id/version. I saw it mentioned as rtl8637b and rtl8637rb. They are different things. rtl8637b is probably just the driver name (like rtl8367c is the one after it). The rtl8365mb.c already supports RTL8367RB-VB, which might be just your switch (or not). rtl8637b driver prints some info about the switch when it detects it but I didn't see any bootlog in this discussion. If the rtl8365mb.c driver does not detect it, you need to add it to rtl8365mb_chip_infos in rtl8365mb.c.

Now the bad news: the rtl8365mb driver does not support forwarding offload nor the vlan feature. Any traffic between ports are forced into the CPU. I worked with a similar situation with the TP-Link Archer C5-v4, that has a SoC switch and uses rtl8367s for external ports. In that case, I needed to disable port learning in the SoC switch to make that work. I do have forwarding and vlan implemented in my tree (https://github.com/luizluca/openwrt/tree/c5v4_dsa), but the maintainer rejected as he was working in an almost ready alternative (that never landed). That branch is also a nice example of migrating a realtek switch from swconfig to dsa if you dare.

Reply:
Thank you so much. I will play a bit and see what i get.

At the moment i am not sure what i am even looking at and am mostly trying to gather bits and pieces from work done by others and spread all over the place.

Here are some bootlogs http://arny.tjps.eu/OpenWrt/EasyBox904xDSL/oem-firmware-info/

and also a document talking about RTL8367RB http://arny.tjps.eu/OpenWrt/EasyBox904xDSL/oem-firmware-info/research/Findings.txt

plus this http://arny.tjps.eu/OpenWrt/EasyBox904xDSL/oem-firmware-info/research/disass_rtl8367rb.ko_mdio-smi.txt

So looks like there is a good amount of information, i just do not understand it yet.

Reply:
@henning-schild if you want to revive support for this board I can share some of my experience.
First advice is: start small and work in steps
1. get serial console working and boot to initramfs
2. basic Ethernet
3. possibly USB (this is typically just a case of enabling the right GPIOs for VBUS)

For basic Ethernet I'd start with the DSL/WAN port.
According to https://github.com/openwrt/openwrt/commit/ac583d9aa63dfcee0132595430a1ba1a45f76ac1 it serves dual purpose, much like on vgv7510kw22
This means:
- two pairs from an Ethernet cable are routed to the VRX200 SoC (PHY addr 0x13, switch port 4)
- another pair is routed to the xDSL modem
- the last pair is unused

I think the setup from https://github.com/openwrt/openwrt/blob/3b221ba5afe42bc32e878299d7d2275702536684/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_netgear_dm200.dts#L91-L108 would give you 10/100 Ethernet on the DSL port.
The first learning here is to differentiate PHYs, addressing on the MDIO bus and the addressing on the switch port (MACs). This is also where I realized that Ethernet isn't as easy as "just plugging in a cable".

VRX200 SoCs have two built-in (so-called) GPHYs.
Each GPHY can be configured to act as one Gbit/s PHY or two 100Mbit/s PHYs.
Those PHYs are using a SoC internal connection - which is why you find `phy-mode = "internal"`.
In the example above the GPHY is configured to "FE" mode (2x 100Mbit/s PHYs - of which one is not connected externally).

Two switches are typically connected with a MAC to MAC connection, there aren't any PHYs between those MACs (MDIO is sometimes still used to communicate with the switch, even if there's no PHY - but in case of the Realtek switch this is not the case AFAIK).
Ethernet PHYs are responsible for detecting whether the link is up (= another device is connected at the other end). Since there's no PHYs in MAC to MAC communication the link speed, status, etc. has to be established otherwise: this is achieved using a `fixed-link` (see `vr9.dtsi` where the Ethernet MAC `&eth0` has a fixed link and so does CPU port on the switch at `port@6`).
As the other switch is an external device the proper `phy-mode` has to be found. As the Realtek switch primary port speed is 1Gbit/s it will be connected via RGMII (beware that RGMII is tricky and requires RX/TX delays to be setup up properly - and to make things trickier, in theory there's three places to do so: MAC, PCB trace lengths and PHY).
Those are the very basics to wire up two MACs together in .dts. Others have already given advice on what to do with the Realtek switch (which I have no experience with).

(I think the wireless iNIC is also using a MAC to MAC connection - but I think it'll be hard to get this supported upstream)